raft/src/main/java/org/apache/kafka/raft/RaftClient.java

```diff
@@ -24,14 +24,37 @@
 public interface RaftClient {
```

Comment:
As a user of this interface am I guaranteed that `listener` and the futures returned by `append` and `read` will be called by the same thread? Or is the user required to deal with the possibility that they are called by different threads?

Reply:
I remember that we had a discussion about callbacks APIs `ElectionListener` vs `CompletableFuture` APIs. I don't have a strong opinion of one over the other but it would be nice if the API used one.

I understand that the issue is that `ElectionListener` is more like an async stream since it can fire multiple times with different values. While `append` and `read` can only return one value.

I can think of two solutions:
1. CompletableFuture<LeaderState> leaderState(LeaderState knownState) - the returned future completes if/when the leader state is different from the known state. The known state can something like `Optional<Int>` for optional epoch. The first time it is called the state machine can assume it is a follower by sending `Optional.empty()`.
2. Use Java's Flow: https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.html

Reply:
I thought a little bit about the `leaderState` api and it just seems awkward to me. It reminds me a bit of how zookeeper watches work. You wouldn't be guaranteed to see all leader changes because you have to be subscribed at the time the event happens. I'm not sure if that matters or not for all state machines, but I know we have tended to get into trouble in zookeeper when we missed some events.

The key issue is probably making the threading assumptions clear. I guess by suggesting the `Flow` API, what you have in mind is for the raft client provide an asynchronous stream of events which can be monitored by the state machine. Then callbacks are not synchronous and we do not depend on any locks shared between the state machine and the client. That does sound attractive to me since synchronous callbacks have been another common source of trouble. Let me give this a little bit of thought.

Reply:
Yeah. The API for 1. is not great because it requires that the state machine re-register every time the leader state changes value. As you mention it is also possible for the state to change from Leader to Follower to Leader and not generate any events. To be able to report every state change asynchronously, we will have to have buffer of those events/changes in case the state machine is not keeping up.

Not sure how important this is assuming that the log stays consistent from the point of view of the state machine. If the log changed because of quickly changing leaders, I am sure that we would report this in the `CompletableFutures` returned by `append` and `read`.

Reply:
Yeah, I'm not sure either. For some state machines, it might actually be nice to fast-forward to the latest state. I guess the nice thing about controlling the implementation is that we can change it if we do the wrong thing 🙂 .

Reply:
I think the ordering guarantee that the listener calls be triggered the same as their corresponding `append` is indeed required though a bit implicit since otherwise, e.g. in the ReplicatedCounter, we should not `uncommitted.set(initialValue);` in the listener blindly if a future listener may be fired first successfully and then the latter listener for an older offset would then revert it?

For `read` though, today this ordering is not guaranteed actually since the listener order is based on the read offset, now when it was triggered. Maybe we should make that clearer in the javadocs.